src/components/lib/dpdk/cos_dpdk.c

```diff
@@ -0,0 +1,504 @@
+			dev_info.min_rx_bufsize,
+			dev_info.max_rx_pktlen,
+			dev_info.max_rx_queues,
+			dev_info.max_tx_queues,
+			dev_info.rx_desc_lim.nb_min,
+			dev_info.rx_desc_lim.nb_max,
+			dev_info.tx_desc_lim.nb_min,
+			dev_info.tx_desc_lim.nb_max,
+			link_status_text
+			);
+
+	return;
+}
+
+/*
+ * cos_eth_ports_init: find and init all ether ports that are available
+ *
+ * @return: nb_ports: 0 no port found
+ * 
+ * note: each port repesents an Ether NIC, and ports whose link status is down will be ignored
+ */
+uint16_t
+cos_eth_ports_init(void)
+{
+	cos_portid_t port_id;
+	uint16_t i;
+
+	memset(ports_ids, 0, sizeof(ports_ids));
+
+	RTE_ETH_FOREACH_DEV(port_id) {
+		ports_ids[nb_ports] = port_id;
+		nb_ports++;
+	}
+
+	COS_DPDK_APP_LOG(NOTICE, "cos_eth_ports_init success, find %d ports\n", nb_ports);
+
+	for(i = 0; i < nb_ports; i++) {
+		cos_eth_info_print(ports_ids[i]);
+	}
+
+	return nb_ports;
+}
+
+static struct rte_mempool * cos_dpdk_pktmbuf_pool = NULL;
+static struct rte_eth_conf default_port_conf = {
+	.rxmode = {
+		.split_hdr_size = 0,
+	},
+	.txmode = {
+		.mq_mode = RTE_ETH_MQ_TX_NONE,
+	},
+};
+
+/*
+ * cos_create_pkt_mbuf: wrapper function for rte_pktmbuf_pool_create
+ *
+ * @name: pkt pool name
+ * @nb_mbufs: number of mbufs within this pool, thus the maximum packets
+ *            that can be stored in this single mem pool.
+ * 
+ * @return: NULL on allocate failure, others on success
+ * 
+ * note: this function assumes that the size of each mbuf is COS_MBUF_DEFAULT_BUF_SIZE.
+ *       this setting should be enough for most use cases and thus is intended to simplify
+ *       users' programming overhead.
+ */
+cos_mbuf_pool_t
+cos_create_pkt_mbuf_pool(const char *name, size_t nb_mbufs)
+{
+	#define MEMPOOL_CACHE_SIZE 256
+	return rte_pktmbuf_pool_create(name, nb_mbufs,
+		MEMPOOL_CACHE_SIZE, 0, COS_MBUF_DEFAULT_BUF_SIZE,
+		rte_socket_id());
+}
+
+/*
+ * cos_free_packet: wrapper function for rte_pktmbuf_pool_create
+ *
+ * @name: pkt pool name
+ * 
+ * @return: NULL on allocate failure, others on success
+ * 
+ * note: this function will free the mbuf used by this packet
+ */
+int
+cos_free_packet(cos_mbuf_t packet)
+{
+	/* TODO: add free logic */
+	return 0;
+}
+
+/*
+ * cos_config_dev_port_queue: wrapper function for rte_eth_dev_configure
+ *
+ * @port_id: eth port id, from user's perspective, the maximum id is get
+ *           from cos_eth_ports_init
+ * @nb_rx_q: number of rx queues used with this port
+ * @nb_tx_q: number of tx queues used with this port
+ * 
+ * return: 0 on success, others will cause panic
+ * 
+ * note: this function gives users ability to config a port's rx/tx queues
+ */
+int
+cos_config_dev_port_queue(cos_portid_t port_id, uint16_t nb_rx_q, uint16_t nb_tx_q)
+{
+	int ret;
+	struct rte_eth_conf local_port_conf = default_port_conf;
+	ret = rte_eth_dev_configure(ports_ids[port_id], nb_rx_q, nb_tx_q, &local_port_conf);
+	if (ret < 0)
+		rte_exit(EXIT_FAILURE, "Cannot configure device: err=%d, port=%u\n",
```

Comment:
If the statement is on a new line, please always `{}`. It really hurts legibility if the condition body is on another line without `{}`.

Reply:
Yes, I forgot it. This is the DPDK style. I will change this.